File: graph/kruskal.py
# https://atcoder.jp/contests/jag2012autumn/tasks/icpc2012autumn_c
# https://github.com/drken1215/book_algorithm_solution/blob/master/codes/chap11/code_11_1.cpp
# https://github.com/drken1215/book_algorithm_solution/blob/master/codes/chap11/code_11_4.cpp
# https://github.com/drken1215/book_algorithm_solution/blob/master/codes/chap15/code_15_1.cpp

import sys

class UnionFind:
    def __init__(self, N):
        self.parent = [-1] * N
        self._size = [1] * N

    def root(self, x):
        if self.parent[x] == -1:
            return x;
        else:
            ans = self.root(self.parent[x])
            self.parent[x] = self.root(self.parent[x])
            return ans

    def issame(self, x, y):
        return self.root(x) == self.root(y)

    def unite(self, x, y):
        x = self.root(x)
        y = self.root(y)
        if x == y:
            return False
        else:
            if self._size[x] < self._size[y]:
                x, y = y, x  # Union by size
            self.parent[y] = x
            self._size[x] += self._size[y]
            return True

    def size(self, x):
        return self._size[self.root(x)]


def main(n, m):
    edges = []
    for i in range(m):
        s, t, c = map(int, sys.stdin.readline().split())
        edges.append((c, s - 1, t - 1))
    edges.sort()
    # Kruskal's algorithm O(|E|log|V|)
    weights = []
    uf = UnionFind(n)
    for w, u, v in edges:
        if uf.issame(u, v):
            continue
        weights.append(w)
        uf.unite(u, v)
    print(weights[n // 2 - 1])        

if __name__ == '__main__':
    while True:
        n, m = map(int, sys.stdin.readline().split())
        if n == 0 and m == 0:
            exit(0)
        main(n, m)

File: graph/kruskal.cc
// https://atcoder.jp/contests/abc065/tasks/arc076_b
// https://github.com/drken1215/book_algorithm_solution/blob/master/codes/chap11/code_11_1.cpp
// https://github.com/drken1215/book_algorithm_solution/blob/master/codes/chap11/code_11_4.cpp
// https://github.com/drken1215/book_algorithm_solution/blob/master/codes/chap15/code_15_1.cpp

#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

struct UnionFind {
  vector<int> parent, _size;

  UnionFind(int n) : parent(n, -1), _size(n, 1) {}

  int root(int x) {
    if (parent[x] == -1) {
      return x;
    } else {
      //return root(parent[x]);
      return parent[x] = root(parent[x]);
    }
  }

  bool issame(int x, int y) {
    return root(x) == root(y);
  }

  bool unite(int x, int y) {
    x = root(x);
    y = root(y);
    if (x == y) {
      return false;
    }
    if (_size[x] < _size[y]) {
      swap(x, y);  // Union by size
    }
    parent[y] = x;
    _size[x] += _size[y];
    return true;
  }

  int size(int x) {
    return _size[root(x)];
  }
};

int main() {
  int N;
  cin >> N;
  vector<pair<int, int> > x(N);
  vector<pair<int, int> > y(N);
  for (int n = 0; n < N; n++) {
    cin >> x[n].first >> y[n].first;
    x[n].second = n;
    y[n].second = n;
  }
  sort(x.begin(), x.end());
  sort(y.begin(), y.end());
  vector<pair<int, pair<int, int> > > edges;
  for (int n = 0; n < N - 1; n++) {
    edges.push_back(make_pair(x[n + 1].first - x[n].first, make_pair(x[n].second, x[n + 1].second)));
    edges.push_back(make_pair(y[n + 1].first - y[n].first, make_pair(y[n].second, y[n + 1].second)));
  }
  // Kruskal's algorithm O(|E|log|V|)
  sort(edges.begin(), edges.end());
  long long weight = 0;
  UnionFind uf(N);
  for (size_t i = 0; i < edges.size(); i++) {
    int w = edges[i].first;
    int u = edges[i].second.first;
    int v = edges[i].second.second;
    if (uf.issame(u, v)) {
      continue;
    }
    weight += w;
    uf.unite(u, v);
  }
  cout << weight << endl;
  return 0;
}
